crates/sui-framework/packages/sui-framework/sources/k_of_n_upgrade_policy.move

```diff
@@ -0,0 +1,511 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Simple upgrade policy that requires a `k` out of `n` quorum in order to perform
+/// a proposed upgrade.
+/// 
+/// This policy is initiated with a call to `k_of_n_upgrade_policy::new` providing 
+/// the `UpgradeCap` of the package to be controlled by the policy, the `k` value 
+/// (number of votes to be received for the upgrade to be allowed) and the list of
+/// `address`es allowed to vote. The `address`es provided will receive
+/// a `Ballot` for any given proposal. The sender of the transaction that creates
+/// the `KofNUpgradeCap` is added to the list of voters as well.
+/// The set of possible voters is provided at creation time and it is immutable for
+/// the lifetime of the policy.
+/// This policy allows for a number of voters between 2 and 100.
+/// 
+/// An upgrade is proposed via `k_of_n_upgrade_policy::propose_upgrade` and the sender
+/// of the upgrade not only needs to have a reference to the `KofNUpgradeCap` but they 
+/// must also be in the list of possible voters.
+/// A successful call for a valid proposal will result in all `address`es
+/// registered to receive a `Ballot` which would allow them to vote for the proposal.
+/// The proposer will not receive a `Ballot`.
+/// Receiver of the `Ballot` can then vote for the upgrade. The `Ballot` can also
+/// be sent to other entities which can then vote for the upgrade on behalf of
```

Comment:
One thing worth mentioning is that this ballot structure provides an interesting incentive structure to voting. In particular, proposing an upgrade is expensive (you are creating many objects), but voting for an upgrade (or destroying your ballot) is either free or incentivized to recoup the storage fee.

This is not an incentive structure that is present in an address-based or muli-use-capability based system.

Reply:
great point, though I have removed the ballot for now and we can add it later.
Other comments expalin...